typing/typecore.ml

```diff
@@ -4833,14 +4833,38 @@ let report_error env ppf = function
       fprintf ppf "Variable %s must occur on both sides of this | pattern"
         (Ident.name id);
       spellcheck_idents ppf id valid_idents
-  | Expr_type_clash (trace, explanation) ->
+  | Expr_type_clash (trace, explanation, exp) ->
       report_unification_error ppf env trace
         ~type_expected_explanation:
           (report_type_expected_explanation_opt explanation)
         (function ppf ->
            fprintf ppf "This expression has type")
         (function ppf ->
-           fprintf ppf "but an expression was expected of type")
+           fprintf ppf "but an expression was expected of type");
+
+      let hints elt =
+        match elt, exp with
+        | Ctype.Unification_trace.(Diff
+            { expected = { t = { desc = Tconstr (p2, [], _); _ }; _ } }),
+          Some (Texp_constant (Const_int n)) ->
+            let hint_suffix c =
+              fprintf ppf "@,@[Hint: Did you mean `%d%c'?@]" n c
+            in
+            if Path.same p2 Predef.path_int32 then
+              hint_suffix 'l'
+            else if Path.same p2 Predef.path_int64 then
+              hint_suffix 'L'
+            else if Path.same p2 Predef.path_nativeint then
+              hint_suffix 'n'
+            else if Path.same p2 Predef.path_float then
+              hint_suffix '.'
+            else
+              ()
+        | _ -> ()
+      in
+
+      List.iter hints trace
```

Comment:
If I am not mistaken, this `List.iter` is only used to find the last element of the trace. I think it might be more readable to add an `explanation` function to `Ctype.Unification_trace` which would extract this last element.

Reply:
You are right. I added the explain function, I took it from `printtyp.ml`.